experimental/alp/lib/Transforms/legalize_vector.cpp

```diff
@@ -98,103 +107,108 @@ struct ForOpVectorProgates : public OpRewritePattern<scf::ForOp> {
     forOp.getResult(0).replaceAllUsesWith(newForOp.getResult(0));
 
     SmallVector<Value, 4> newResults;
-    for (unsigned i =0; i<forOp.getNumResults(); i++){
+    for (unsigned i = 0; i < forOp.getNumResults(); i++) {
       newResults.push_back(newForOp.getResult(i));
     }
     rewriter.replaceOp(forOp, newResults);
     // Get the users of the original result
     auto users = newForOp.getResult(0).getUsers();
 
     // And now let's change this
-    for (Operation *user : users){
-      if (auto extractOp = dyn_cast<ExtractStridedSliceOp>(user)){
-          int64_t offset1 = extractFromI64ArrayAttr(extractOp.offsets())[0];
-          int64_t offset2 = extractFromI64ArrayAttr(extractOp.offsets())[1];
-          if (offset1 == working_offset1 && offset2 == working_offset2){
-              extractOp.getResult().replaceAllUsesWith(newForOp.getResults().back());
-              rewriter.eraseOp(extractOp);
-          }
+    for (Operation *user : users) {
+      if (auto extractOp = dyn_cast<ExtractStridedSliceOp>(user)) {
+        int64_t offset1 = extractFromI64ArrayAttr(extractOp.offsets())[0];
+        int64_t offset2 = extractFromI64ArrayAttr(extractOp.offsets())[1];
+        if (offset1 == working_offset1 && offset2 == working_offset2) {
+          extractOp.getResult().replaceAllUsesWith(
+              newForOp.getResults().back());
+          rewriter.eraseOp(extractOp);
+        }
       }
     }
     return success();
   }
 };
 
-struct LegalizePass : public LegalizePassBase<LegalizePass>{
+struct LegalizePass : public LegalizePassBase<LegalizePass> {
   LegalizePass() = default;
-  LegalizePass(const LegalizePass&pass) {}
+  LegalizePass(const LegalizePass &pass) {}
 
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<AffineDialect>();
   }
 
-  void runOnFunction() override {
+  void runOnOperation() override {
     auto *ctx = &getContext();
-    if (getFunction().getName() != "kernel"){
-        return;
+    if (getOperation().getName() != "kernel") {
+      return;
     }
+    LLVM_DEBUG(llvm::dbgs() << "legalize starts\n ");
+
     RewritePatternSet patterns(ctx);
     populateVectorUnrollPatterns(
         patterns,
         UnrollVectorOptions().setNativeShapeFn(getShape).setFilterConstraint(
             filter));
-    (void)applyPatternsAndFoldGreedily(getFunction(), std::move(patterns));
+    (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
 
     // Now there will be a lot of:
     // - transfer_read/insert_strided_slices
     // - transfer_write/extract_strided_slices
-    // Next step is to remove as many {insert,extract}_strided_slices as we can, especially across loop blocks
+    // Next step is to remove as many {insert,extract}_strided_slices as we can,
```

Comment:
This feels like it should be covered by union of canonicalization and folding patterns.
Could you share examples of IR before and after applying this transformations?
This way we can track down what is missing here.

Reply:
I added a test in `experimental/alp/test/test-alp-legalize.mlir`. See if you are able to disable the additional pattern and still don't end up with lots of insert/extract slices, I tried hard but wasn't able to do that

Reply:
ok, will take a look separately, thanks!